# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/content/widgets/browser-custom-element.js
# Commit: 30d2bf04719b
# Full Hash: 30d2bf04719bfb60095aaa7d679363e57d25ad92
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1640019
# File Overlap Count: 1
# Description:
#   Bug 1640019 - Part 1: Support toplevel process switches outside of tabbrowser, r=mattwoodrow
#   
#   This new process switching behavior is only enabled for some browser elements,
#   which have specified a specific attribute. Turning this on for all browsers with
#   a `remote` attribute causes breakage in reftests.
# ==============================================================================

diff -r 3fefa1d023a6 -r 30d2bf04719b toolkit/content/widgets/browser-custom-element.js
--- a/toolkit/content/widgets/browser-custom-element.js	Fri Jun 12 17:02:45 2020 +0000
+++ b/toolkit/content/widgets/browser-custom-element.js	Fri Jun 12 16:52:00 2020 +0000
@@ -1946,8 +1946,29 @@
       return origins;
     }
 
-    get canPerformProcessSwitch() {
-      return this.getTabBrowser() != null;
+    get processSwitchBehavior() {
+      // If a `remotenessChangeHandler` is attached to this browser, it supports
+      // having its toplevel process switched dynamically in response to
+      // navigations.
+      if (this.hasAttribute("maychangeremoteness")) {
+        return Ci.nsIBrowser.PROCESS_BEHAVIOR_STANDARD;
+      }
+
+      // When loaded in a TabBrowser, use the custom behavior from
+      // `performProcessSwitch`.
+      if (this.getTabBrowser() != null) {
+        return Ci.nsIBrowser.PROCESS_BEHAVIOR_CUSTOM;
+      }
+
+      // For backwards compatibility, we need to mark remote, but
+      // non-`allowremote`, frames as `PROCESS_BEHAVIOR_SUBFRAME_ONLY`, as some
+      // tests rely on it.
+      // FIXME: Remove this?
+      if (this.isRemoteBrowser) {
+        return Ci.nsIBrowser.PROCESS_BEHAVIOR_SUBFRAME_ONLY;
+      }
+      // Otherwise, don't allow gecko-initiated toplevel process switches.
+      return Ci.nsIBrowser.PROCESS_BEHAVIOR_DISABLED;
     }
 
     performProcessSwitch(
